appcelerator/titanium_mobile
# fix(android): Refactored activity/fragment restore handling

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26966
- https://jira.appcelerator.org/browse/TIMOB-17089
- https://jira.appcelerator.org/browse/TIMOB-26890
- https://jira.appcelerator.org/browse/TIMOB-26964
- https://jira.appcelerator.org/browse/TIMOB-26914
- https://jira.appcelerator.org/browse/TIMOB-26975
- https://jira.appcelerator.org/browse/TIMOB-26978
- https://jira.appcelerator.org/browse/TIMOB-15829
- https://jira.appcelerator.org/browse/TIMOB-26989
- https://jira.appcelerator.org/browse/TIMOB-26996

**Summary:**
- [TIMOB-26966](https://jira.appcelerator.org/browse/TIMOB-26966) Fixed bug where back navigating from child window causes app to exit when "Don't keep activities" is enabled as of 7.2.1.
- [TIMOB-26914](https://jira.appcelerator.org/browse/TIMOB-26914) Fixed 8.0.0 regression where an OS forced quit app displays a blank window upon relaunch. (Was incorrectly restoring child activity.)
- [TIMOB-17089](https://jira.appcelerator.org/browse/TIMOB-17089) Resolved view ID conflicts causes app crashes when activity/fragment is restored.
- [TIMOB-26964](https://jira.appcelerator.org/browse/TIMOB-26964) Fixed bug where TabGroup would be blank (no tabs) when calling child window's close() method while "Don't keep activities" is enabled.
- [TIMOB-26890](https://jira.appcelerator.org/browse/TIMOB-26890) Fixed bug where tapping 4th tab in TabGroup can cause a crash due to view ID conflict during fragment restore.
- [TIMOB-26975](https://jira.appcelerator.org/browse/TIMOB-26975) Fixed bug where LiveView does not restart app if "Don't keep activities" is enabled as of 8.0.0.
- [TIMOB-26978](https://jira.appcelerator.org/browse/TIMOB-26978) Added several safety mechanisms to `AlertDialog` and `ProgressIndicator` dialog code to avoid crashes when activity is destroyed, but not finished. (Can happen when "Don't keep activities" is enabled.)
- [TIMOB-15829](https://jira.appcelerator.org/browse/TIMOB-15829) Fixed crash caused by progress indicator set up on the status bar if shown after hiding twice.
- [TIMOB-26989](https://jira.appcelerator.org/browse/TIMOB-26989) Fixed crash caused by restoring an activity having a `TableView` with a `SearchBar` or `SearchView` attached to it. (Proxies should not be released unless activity is finished.)
- [TIMOB-26996](https://jira.appcelerator.org/browse/TIMOB-26996) Fixed bug where Window/TabGroup `close()` method call was ignored until "open" event was fired.
- Increased TabGroup's ViewPager offscreen page limit from 1 to 128 to avoid tab fragments from being destroyed/restored.

---
**Test Restore After OS Force Quit:**
1. On Android, go to its "Developer Options" screen.
2. Tap on the "Background process limit" row. (It's near the bottom of the list.)
3. Tap on "No background processes". _(This tells the OS to force quit the last displayed app when launching another app.)_
4. Build and run the below code on the same Android device.
5. In the app, tap on the "Show Child Window" button.
6. Press the Android "Home" button.
7. Tap on another app such as "Gmail".
8. Press the Android "Home" button.
9. Resume the Titanium app.
10. Verify that the app restarts, starting with the splash screen. _(It will launch with a blank window, but quickly restart. Unfortunately, we are stuck with this behavior.)_
11. Change device's "Background process limit" setting back to "Standard".

```javascript
var window = Ti.UI.createWindow();
var openButton = Ti.UI.createButton({ title: "Show Child Window" });
openButton.addEventListener("click", function(e) {
	var childWindow = Ti.UI.createWindow({ backgroundColor: "orange" });
	var closeButton = Ti.UI.createButton({ title: "Close Me" });
	closeButton.addEventListener("click", function(e) {
		childWindow.close();
	});
	childWindow.add(closeButton);
	childWindow.open();
});
window.add(openButton);
window.open();
```

---
**Test Intent ACTION_VIEW with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Tap the "Send Intent" button.
7. Verify that a screenshot of the "Send Intent" button is shown in a new window.
8. Tap the Android "Home" button.
9. Resume the app.
10. Verify that the screenshot is still shown.
11. Tap the Android "Back" button.
12. Verify that the it navigated back to the Titanium app with the "Send Intent" button. _(Should no longer exit the app, which was the bug.)_
13. Tap the Android "Back" button.
14. Verify that it exits the app. You should not see a splash screen.

```javascript
var window = Ti.UI.createWindow();
var button = Ti.UI.createButton({
	title: "Send Intent",
});
button.addEventListener("click", function(e){
	button.toImage(function(imageBlob) {
		if (!imageBlob) {
			return;
		}
		var targetFile = Ti.Filesystem.getFile(Ti.Filesystem.applicationDataDirectory, "image.png");
		targetFile.write(imageBlob, false);
		var intent = Ti.Android.createIntent({
			action: Ti.Android.ACTION_VIEW,
			data: targetFile.nativePath,
			type: "image/png",
		});
		intent = Ti.Android.createIntentChooser(intent, "Choose");
		Ti.Android.currentActivity.startActivity(intent);
	});
});
window.add(button);
window.open();
```

---
**Test TabGroup with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Tap on the "Open Child Window" button.
7. Tap on the "Close Window" button.
8. Verify that you are returned back to a 3 tab window. _(Should not be blank, which was the bug.)_

```javascript
function createTab(title) {
	var window = Ti.UI.createWindow({ title: title });
	window.add(Ti.UI.createLabel({
		text: title + " View",
		top: "25%",
	}));
	var openButton = Ti.UI.createButton({
		title: "Open Child Window",
		top: "50%",
	});
	openButton.addEventListener("click", function() {
		var childWindow = Ti.UI.createWindow({
			title: title + " Child Window",
			backgroundColor: "white",
		});
		var closeButton = Ti.UI.createButton({
			title: "Close Window",
		});
		closeButton.addEventListener("click", function() {
			childWindow.close();
		});
		childWindow.add(closeButton);
		childWindow.open();
	});
	window.add(openButton);
	var tab = Ti.UI.createTab({
		title: title,
		window: window,
	});
	return tab;
}

var tabGroup = Ti.UI.createTabGroup({
	title: "TabGroup Parent Test",
	tabs: [createTab("Tab 1"), createTab("Tab 2"), createTab("Tab 3")],
});
tabGroup.open();
```

---
**Test LiveView Restarts with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Edit the "app.js" file and save your changes.
7. Verify that the app restarts and that the below `Ti.API.info()` message is logged again.

```javascript
Ti.API.info("### The 'app.js' file was loaded.");
var window = Ti.UI.createWindow();
window.open();
```
